utils/frame/benchmarking-cli/src/storage/write.rs

```diff
@@ -62,49 +68,102 @@ impl StorageCmd {
-				for (mut k, (_, rc)) in stx.drain().into_iter() {
-					if rc > 0 {
-						db.sanitize_key(&mut k);
-						if db.get(state_col, &k).is_some() {
-							trace!("Benchmark-store key creation: Key collision detected, retry");
-							continue 'retry
-						}
-					}
+				match check_new_value::<Block>(
+					db.clone(),
+					&trie,
+					&k.to_vec(),
+					&new_v,
+					self.state_version(),
+					state_col,
+					None,
+				) {
+					true => {
+						*original_v = new_v;
+						break
+					},
+					false => continue 'retry,
 				}
-				*original_v = new_v;
-				break
 			}
 		}
 
 		info!("Writing {} keys", kvs.len());
 		// Write each value in one commit.
 		for (k, new_v) in kvs.iter() {
-			// Interesting part here:
-			let start = Instant::now();
-			// Create a TX that will modify the Trie in the DB and
-			// calculate the root hash of the Trie after the modification.
-			let replace = vec![(k.as_ref(), Some(new_v.as_ref()))];
-			let (_, stx) = trie.storage_root(replace.iter().cloned(), self.state_version());
-			// Only the keep the insertions, since we do not want to benchmark pruning.
-			let tx = convert_tx::<Block>(db.clone(), stx.clone(), false, state_col);
-			db.commit(tx).map_err(|e| format!("Writing to the Database: {}", e))?;
-			record.append(new_v.len(), start.elapsed())?;
-
-			// Now undo the changes by removing what was added.
-			let tx = convert_tx::<Block>(db.clone(), stx.clone(), true, state_col);
-			db.commit(tx).map_err(|e| format!("Writing to the Database: {}", e))?;
+			let res = calculate_bench::<Block>(
+				db.clone(),
+				&trie,
+				k.to_vec(),
+				new_v.to_vec(),
+				self.state_version(),
+				state_col,
+				None,
+			)?;
+			record.append(res.0, res.1)?;
+		}
+
+		if self.params.include_child {
+			info!("Writing {} child trees", child_tries.len());
+
+			for key in child_tries {
+				let trie_id = key
+					.strip_prefix(DEFAULT_CHILD_STORAGE_KEY_PREFIX)
+					.expect("Checked above to exist");
+				let info = ChildInfo::new_default(trie_id);
+				let child_keys =
+					client.child_storage_keys_iter(&block, info.clone(), None, None)?;
+
+				for ck in child_keys {
+					if let Some(original_v) = client
+						.child_storage(&block, &info.clone(), &ck)
+						.expect("Checked above to exist")
+					{
+						let mut new_v = vec![0; original_v.0.len()];
+						'child_retry: loop {
+							rng.fill_bytes(&mut new_v[..]);
+							match check_new_value::<Block>(
+								db.clone(),
+								&trie,
+								&ck.0,
+								&new_v,
+								self.state_version(),
+								state_col,
+								Some(&info),
+							) {
+								true => break,
+								false => continue 'child_retry,
+							}
+						}
+
+						let res = calculate_bench::<Block>(
+							db.clone(),
+							&trie,
+							ck.0,
+							new_v.to_vec(),
+							self.state_version(),
+							state_col,
+							Some(&info),
+						)?;
+						record.append(res.0, res.1)?;
```

Comment:
In the future we can split up the results into a `top_record` and `child_record` and then have two resulting weight constants.  
Then the developer can decide how much he wants to use them. But for now it should be good like this.

Reply:
I agree, separating average for child tree nodes and main tree nodes makes sense.

Reply:
Would have to think a bit which solution is best, so let's do it afterwards https://github.com/paritytech/substrate/issues/12046